Make a proper Log class and methods

plocket opened the pull request:
Apparently there was still a while to go. I still have decision and design/architecture docs to write, but this is ready for review. What I most want to discuss is how to handle bugs internal to `.debug()`.

I wish this integrated the report more too, but this is already big enough.

In this PR, I have:

* [x] Added tests for any new features or bug fixes (if relevant)
* [x] Added my changes to the CHANGELOG.md at the top, under the "Unreleased" section
* [x] Ensured issues that this PR closes will be [automatically closed](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue)

### Reason for this PR

The logs have been a bit of a mess for a while. The verbose logs could only be logged live, not saved. Saving to the debug log was a bit add-hoc. There wasn't a good differentiation between console logs and debug logs. This addresses a few of those problems.

### Links to any solved or related issues

Close #805, create verbose-type log
Close #924, remove extra url print
Close #659, abstract adding to debug_log
Close #925, allow a `Log` to throw an error
Address #461, setup and takedown reports

### Any manual testing I have done to ensure my PR is working

Visually checked various log files.

plocket committed:
Brainstorm log goals and architecture

Co-authored-by: will-morningstar <mxwillmorningstar@gmail.com>

plocket committed:
Implement draft of log.verbose() and

- Put a wait in time.js
- Modified review doc based on more findings

plocket committed:
Implement and lightly test basic log methods

plocket committed:
Start replacing `log` with `Log`

plocket committed:
Append to files, add error checking and fallbacks

Co-authored-by: Russ741 <72242656+Russ741@users.noreply.github.com>

plocket committed:
Untested: make artifacts folder, add info logs, also

- Ask questions
- Experiment with missing logs on error. See https://replit.com/@plocket/Avoid-leaving-a-consoleConsole-file-blank-on-error#index.js
- Describe new log-calling signature
- Re-throw errors
- Preserve log levels in debug console
- Use `recursive` when creating artifacts folder

Co-authored-by: will-morningstar <mxwillmorningstar@gmail.com>
Co-authored-by: cthulahoops <adam@cthulahoops.org>

plocket committed:
Update decisions, alias waiting, tweak logs

plocket committed:
Implement new log interface for setup, run, takedown, da api,

Also, update codes

Co-authored-by: cthulahoops <adam@cthulahoops.org>

plocket committed:
Implement new log signature in all necessary files
Also:
* Improve some logs
* Implement logs in unit tests

plocket committed:
Add logs decision doc, remove success rainbow from action

plocket committed:
Change log methods signature, use fs.append, add stdout...

- Add report_log.txt
- Implement action output (untested)
- Implement logs in unit tests

Co-authored-by: will-morningstar <mxwillmorningstar@gmail.com>

plocket committed:
Fix workflow typos, remove ALK0000, add TODOs, edit decision docs

plocket committed:
Add logging to actions

plocket committed:
Create folder and paths for log tests, file creation handles path

Also tweak action log

plocket committed:
Brainstorm and refactor to do calcs in .debug(). Also,

- .console and .debug signatures diverge. Reconsider.
- Always throw when calling `.throw`

TODO: When we throw `null` we seem to not get a call stack. We
gotta do something about that. Maybe always make an Error if
`do_throw` is true

Co-authored-by: will-morningstar <mxwillmorningstar@gmail.com>

plocket committed:
Refactor formatting, add tests, clean up

Co-authored-by: will-morningstar <mxwillmorningstar@gmail.com>

plocket committed:
Brainstorm about categories of errors, add notes, refactor, add

tests, commit before error refactor

Co-authored-by: will-morningstar <mxwillmorningstar@gmail.com>

plocket committed:
Remove explicitly writing to unexpected, note about writing to

reports, warn in debug on error, catch more.

Note: writing to debug will write to unexpected

plocket committed:
Only allow `error` on throw, warn of misuse, make

writing to unexpected file private, refactor to remove extra try/catches, match up log codes

Tests: Try to neutralize console. Unable. Add new tests, add check string exclusion.

Add notes

plocket committed:
Join stdout args with no spaces. Brainstorm class signature with

@ethanstrominger

Co-authored-by: ethanstrominger <ethanstrominger2@gmail.com>

plocket committed:
Tweak name, add function descriptions, comments, and examples

BryceStevenWilley commented on docs/decisions/006_which_log_level_to_pick_2024_09_5.md:
```diff
@@ -0,0 +1,46 @@
+`error`:
+  - Avoid using `error` explicitly. Only the debug logs should use the `error` level, and only to record critical errors that come in with `throw`.
+`info`:
+  - Log expected errors as `info`. For example, the aborted promises when awaiting navigation.
```
Really interesting rule for what to use when. Trying to sit with my biases of what INFO is usually used for, and determine if it's worth renaming INFO to something else, like "suspicious" (SUS). Something that, by itself, can be ignored (and doesn't need to be investigated), but with a failure, might be an issue. Usually I'd use warn for that, but I do like warn being only for user focused logs

plocket replied:
I like this line of thought. No need to reply if you don't have bandwidth, I'll just brainstorm before.

Non-alarming words that can have an internal meaning:
- note
- detail
- cue
- queue
- data
- trace
- dig
- finding
- memo
- review
- intel
- mark
- comms
- signal
- task
- todo

Somewhat alarming words that can have an internal meaning:
- suspicious
- sus
- alert
- notice
- clue
- mark
- symptom
- advise
- caution

We could also require using "info" for things that don't need attention and leave "debug" as our internal signal to examine further. Or maybe, going in a different direction - just "alkiln"/"alk".

plocket replied:
Misread this note a bit. I think an alternative word (like "SUS") would be useful for internal warnings, but just want to be clear that an aborted promise is expected and is really just information - when one promise wins a race, the others will, of course, be (correctly) aborted. I added a design doc about the various meanings of "error" that might help clarify this a bit. I sort of want to combine that doc with this one, but I haven't seen how to make it clear.

BryceStevenWilley commented on lib/docassemble/server_install.js:
```diff
@@ -23,20 +24,33 @@ const server_install = async () => {
-      code: `ALK0045`,
-      data: `There are ${folders.length} folders at ${path}: "${folders.join('", "')}". ALKiln chose to use the folder "${chosen_folder}".`
-    })
+    log.warn({ code: `ALK0045`, context: `setup`, },
```
Still working through things, but should `context` be changed to `server_install` too?
```suggestion
    log.warn({ code: `ALK0045`, context: `server_install`, },
```

plocket committed:
Refactor getting errors, protect against infinite loops...

- Ensure objects don't always log with a leading new line
- Unify language around caught behavior/misbehavior to "Skipped"
- Protect stdout
- Tweak file descriptions (thanks Ethan!)

Co-authored-by: ethanstrominger <ethanstrominger2@gmail.com>

BryceStevenWilley commented on docs/decisions/reports_and_logs_2024_08_11.md:
````diff
@@ -0,0 +1,141 @@
+### System-context diagram
+
+```
+logic code -> logger -> report -----------> console
````
I get what these diagrams are trying to show, but why are there 3 different ones? Are those options considered?

plocket replied:
Yes, alternatives. I'm going to have to make some adjustments to this doc and maybe break it out into separate docs.

BryceStevenWilley commented on lib/docassemble/docassemble_api_interface.js:
```diff
@@ -61,6 +61,10 @@ const time = require('../utils/time');
 
+const artifacts_path = session_vars.get_artifacts_path_name();
+const log = new Log({ path: artifacts_path, context: `api` });
+const context = `da api`;
```
```suggestion
```

Unused variable

plocket replied:
Took the other avenue suggested in https://github.com/SuffolkLITLab/ALKiln/pull/948/files#diff-227ae832372018230aee8c35c6536e1bbfbf40bd225b35079529a93f6139713e.

BryceStevenWilley commented on lib/docassemble/docassemble_api_interface.js:
```diff
@@ -61,6 +61,10 @@ const time = require('../utils/time');
 module.exports = da_i;
 
+const artifacts_path = session_vars.get_artifacts_path_name();
+const log = new Log({ path: artifacts_path, context: `api` });
```
(haven't gotten through log.js yet, but) `log` is always called with `context: "da api"` in this file; should `context` here just be `da api` here, and that param can be removed from each individual call?

plocket replied:
I think I understand what you mean and I like it. Though I wish I could make it all caps so it could look like a proper constant. That would somewhat defeat the purpose, though.

BryceStevenWilley commented on docs/decisions/reports_and_logs_2024_08_11.md:
```diff
@@ -0,0 +1,141 @@
+## Non/anti-goals
+
+- Spend a month on this
+- Create a complex and heavy system to try to abstract absolutely everything
```
An additional non-goal that I think you might have; logs that are only understandable if you read the surrounding code. A lot of things can be simplified if you assume that, but it def raises the effort level needed by users quite a bit

plocket replied:
Ongoing struggle to improve clarity, I think. I've (unreasonably optimistically) made a [discussion category for improving quality](https://github.com/SuffolkLITLab/ALKiln/discussions/categories/improving-clarity) for avoiding making tons of issues while getting to put down thoughts about clarifying logs, other output, and maybe documentation (though we do have the repo for the [Assembly Line documentation website](https://assemblyline.suffolklitlab.org/docs/) for that last one probably).

BryceStevenWilley reviewed (commented):
Some initial thoughts. Didn't get a chance to look through everything, but focused on the design docs and how the new log API is used in Kiln itself.

plocket committed:
Refactor deep methods for getting safe error

plocket commented:
> Didn't get a chance to look through everything, but focused on the design docs and how the new log API is used in Kiln itself.

Don't worry about looking at the code implementation for now. I'm doing a fair amount of refactoring. Thanks, @BryceStevenWilley!

plocket committed:
Save both report and debug info to the unexpected results file

plocket committed:
Simplify use of `context` in wrapper functionality as per review

Co-authored-by: BryceStevenWilley <Bryce.Steven.Willey@gmail.com>

plocket committed:
Add and edit decision docs, notes, and TODOs

* Add decision doc for getting and saving logs

Thanks, @jder, for some extra experimentation with
`console.Console` and write streams.

* Describe the 4 environments in which ALKiln runs, tweak function
descriptions in related files.
* Update Log notes and descriptions
* Tweak and clean up Log descriptions/comments, add TODOs
* Focus log level choice doc on log levels instead of Log
* Clarify types of errors and what to do with them. Dedup some info
* Add more notes and content about logs

Co-authored-by: jder <me@jesserusak.com>

plocket commented:
Ready, barring 1 main discussion point (and the docs). That one item - how to handle bugs internal to `.debug()`. They bring up the potential for infinite loops under very specific circumstances. The `max_loops` variable is a safety valve right now, but I hate it.

Thoughts:
1. I really don't like using `max_loops` or some other kind of cut-off. If we use it, we need to make it high enough that useful information doesn't get missed. That said, it just smells bad. Calling `.debug()` internally is just asking for trouble, even if we're calling it with what we believe to be safe arguments.
2. Option - collect the unexpected behavior messages as a list and, at the end, decide whether to save them to the debug log or to console log them. I'm not sure how to pass that data around, though.
3. Option - console log all unexpected behavior. I'm really reluctant to create that much noise for the authors when it's not-critical functionality.

While we mull over that, I'm going to keep plugging away at the design/decision docs.

plocket committed:
Abstract util.inspect()

plocket commented:
My current experiments to eliminate inner calls to `.debug()` is to replace that call to use a function that abstracts saving straight to the debug file.

plocket committed:
Refactor to avoid internal circular calls of `.debug()`

plocket committed:
Implement some TODOs and clean up

plocket committed:
Make internal-specific icon and level, as per PR comment:

https://github.com/SuffolkLITLab/ALKiln/pull/948#discussion_r1802030164

Co-authored-by: BryceStevenWilley <Bryce.Steven.Willey@gmail.com>

plocket committed:
Reformat and deduplicate reports, clarify var names

plocket committed:
Continue work on documentation of logs architecture

* Continue filling documentation
* Describe flow for artifact folder names in 4 environments
* Clarify various docs

plocket committed:
Rename doc about kinds of errors

plocket committed:
Modify artifact folder name diagrams, clarify diagrams

plocket committed:
Add to CHANGELOG, correct/add diagrams. Mermaid needs "o" hack.

plocket committed:
Rename log.js to Log.js

plocket committed:
Merge branch 'v5' into 805_verbose

plocket committed:
Bump "logs" feature release for GitHub tests
